File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
n, m, v = map(int, input().split())

board = {x: [] for x in range(1, n + 1)}

for i in range(m):
    x, y = map(int, input().split())
    board[x].append(y)
    board[y].append(x)


def dfs(board, v):
    need_visited, visited = list(), list()
    need_visited.append(v)
    while need_visited:
        node = need_visited.pop()
        if node not in visited:
            visited.append(node)
            need_visited.extend(sorted(board[node], reverse=True))
    return visited


def bfs(board, v):
    need_visited, visited = list(), list()
    need_visited.append(v)
    while need_visited:
        node = need_visited[0]
        del need_visited[0]
        if node not in visited:
            visited.append(node)
            need_visited.extend(sorted(board[node]))
    return visited


print(" ".join([str(x) for x in dfs(board, v)]))
print(" ".join([str(x) for x in bfs(board, v)]))

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.cc
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <algorithm>

using namespace std;

vector<int> dfs(const map<int, vector<int>>& board, int v) {
    vector<int> visited, need_visited;
    need_visited.push_back(v);
    while (!need_visited.empty()) {
        int node = need_visited.back();
        need_visited.pop_back();
        if (find(visited.begin(), visited.end(), node) == visited.end()) {
            visited.push_back(node);
            if (board.count(node)) {
                vector<int> neighbors = board.at(node);
                sort(neighbors.rbegin(), neighbors.rend());
                need_visited.insert(need_visited.end(), neighbors.begin(), neighbors.end());
            }
        }
    }
    return visited;
}

vector<int> bfs(const map<int, vector<int>>& board, int v) {
    vector<int> visited, need_visited;
    need_visited.push_back(v);
    while (!need_visited.empty()) {
        int node = need_visited.front();
        need_visited.erase(need_visited.begin());
        if (find(visited.begin(), visited.end(), node) == visited.end()) {
            visited.push_back(node);
            if (board.count(node)) {
                vector<int> neighbors = board.at(node);
                sort(neighbors.begin(), neighbors.end());
                need_visited.insert(need_visited.end(), neighbors.begin(), neighbors.end());
            }
        }
    }
    return visited;
}

int main() {
    int n, m, v;
    cin >> n >> m >> v;

    map<int, vector<int>> board;

    for (int i = 0; i < m; ++i) {
        int x, y;
        cin >> x >> y;
        board[x].push_back(y);
        board[y].push_back(x);
    }

    vector<int> dfsResult = dfs(board, v);
    vector<int> bfsResult = bfs(board, v);

    for (int i : dfsResult) {
        cout << i << " ";
    }
    cout << endl;

    for (int i : bfsResult) {
        cout << i << " ";
    }
    cout << endl;

    return 0;
}